File: max_area_of_island.py
import heapq
class Solution(object):

    def maxAreaOfIsland(self,grid):
        """
        : type grid :  List[List[Int]]
        : rtype: Int
        """
        m,n = len(grid),len(grid[0])
        seen = set()
        def dfs(i,j):
            if(0 <= i < m and 0 <= j < n and (i,j) not in seen and grid[i][j]):
                seen.add((i,j))
                return 1 + dfs(i+1,j) + dfs(i-1,j) + dfs(i,j-1) + dfs(i,j+1)
            return 0
        

        areas = [dfs(i,j) for i in range(m) for j in range(n) if grid[i][j]]
        return max(areas) if areas else 0

def main():
    sol = Solution()
    grid = [[0,0,1,0,0,0,0,1,0,0,0,0,0],
            [0,0,0,0,0,0,0,1,1,1,0,0,0],
            [0,1,1,0,1,0,0,0,0,0,0,0,0],
            [0,1,0,0,1,1,0,0,1,0,1,0,0],
            [0,1,0,0,1,1,0,0,1,1,1,0,0],
            [0,0,0,0,0,0,0,0,0,0,1,0,0],
            [0,0,0,0,0,0,0,1,1,1,0,0,0],
            [0,0,0,0,0,0,0,1,1,0,0,0,0]]
    print sol.maxAreaOfIsland(grid)

if __name__ == "__main__":
    main()

File: max-area-of-island.cpp
#include <iostream>
#include <vector>

using namespace std;
class Solution {
public:
  int maxAreaOfIsland(vector<vector<int>>& grid) {
    int column = 0;
    int row = 0;
    int neighbours = 0;
    int count = 0;


    for(vector<vector<int>>::iterator itr = grid.begin();
	itr!=grid.end();
	itr++) {
      for(vector<int>::iterator itr_row = itr->begin();
	  (itr_row)!=itr->end();
	  (itr_row)++) {
	if((*itr_row)==1) {
	  count+=1;
	  if(row!=0 and grid[row-1][column]==1)
	    neighbours+=1;
	  if(column!=0 and grid[row][column-1]==1)
	    neighbours+=1;
	  if((row!=0 and column!=0) and
	     ((grid[row-1][column]==0) and
	      (grid[row][column-1]==0))) {
	    neighbours=0;
	  }
	}
	count = std::max(count,neighbours);
	column+=1;
      }
      column=0;
      row+=1;
    }
    return count;
  }
  
};


